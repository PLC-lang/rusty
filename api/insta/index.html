<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="What are snapshot tests"><title>insta - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceSerif4-Regular-46f98efaafac5295.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/FiraSans-Regular-018c141bf0843ffd.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/FiraSans-Medium-8f9a781e4970d388.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceSerif4-Bold-a2c9cd1067f8b328.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2"><link rel="stylesheet" href="../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../static.files/rustdoc-ba5701c5741a7b69.css" id="mainThemeStyle"><div id="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="insta" data-themes="" data-resource-suffix="" data-rustdoc-version="1.70.0 (90c541806 2023-05-31)" data-search-js="search-e077946657036a58.js" data-settings-js="settings-298e1ea74db45b39.js" data-settings-css="settings-7bfb4c59cc6bc502.css" data-theme-light-css="light-0f8c037637f9eb3e.css" data-theme-dark-css="dark-1097f8e92a01e3cf.css" data-theme-ayu-css="ayu-614652228113ac93.css" ></div><script src="../static.files/storage-62ce34ea385b278a.js"></script><script defer src="../crates.js"></script><script defer src="../static.files/main-f61008743c98d196.js"></script><noscript><link rel="stylesheet" media="(prefers-color-scheme:light)" href="../static.files/light-0f8c037637f9eb3e.css"><link rel="stylesheet" media="(prefers-color-scheme:dark)" href="../static.files/dark-1097f8e92a01e3cf.css"><link rel="stylesheet" href="../static.files/noscript-13285aec31fa243e.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-16x16-8b506e7a72182f1c.png"><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button><a class="logo-container" href="../insta/index.html"><img class="rust-logo" src="../static.files/rust-logo-151179464ae7ed46.svg" alt="logo"></a><h2></h2></nav><nav class="sidebar"><a class="logo-container" href="../insta/index.html"><img class="rust-logo" src="../static.files/rust-logo-151179464ae7ed46.svg" alt="logo"></a><h2 class="location"><a href="#">Crate insta</a></h2><div class="sidebar-elems"><ul class="block"><li class="version">Version 1.34.0</li><li><a id="all-types" href="all.html">All Items</a></li></ul><section><ul class="block"><li><a href="#modules">Modules</a></li><li><a href="#macros">Macros</a></li><li><a href="#structs">Structs</a></li></ul></section></div></nav><main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" title="help" tabindex="-1"><a href="../help.html">?</a></div><div id="settings-menu" tabindex="-1"><a href="../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../static.files/wheel-7b819b6101059cd0.svg"></a></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1>Crate <a class="mod" href="#">insta</a><button id="copy-path" title="Copy item path to clipboard"><img src="../static.files/clipboard-7571035ce49a181d.svg" width="19" height="18" alt="Copy item path"></button></h1><span class="out-of-band"><a class="srclink" href="../src/insta/lib.rs.html#1-326">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><div align="center">
 <img src="https://github.com/mitsuhiko/insta/blob/master/assets/logo.png?raw=true" width="250" height="250">
 <p><strong>insta: a snapshot testing library for Rust</strong></p>
</div>
<h2 id="what-are-snapshot-tests"><a href="#what-are-snapshot-tests">What are snapshot tests</a></h2>
<p>Snapshots tests (also sometimes called approval tests) are tests that
assert values against a reference value (the snapshot).  This is similar
to how <code>assert_eq!</code> lets you compare a value against a reference value but
unlike simple string assertions, snapshot tests let you test against complex
values and come with comprehensive tools to review changes.</p>
<p>Snapshot tests are particularly useful if your reference values are very
large or change often.</p>
<h2 id="what-it-looks-like"><a href="#what-it-looks-like">What it looks like:</a></h2>
<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="attr">#[test]
</span><span class="kw">fn </span>test_hello_world() {
    <span class="macro">insta::assert_debug_snapshot!</span>(<span class="macro">vec!</span>[<span class="number">1</span>, <span class="number">2</span>, <span class="number">3</span>]);
}</code></pre></div>
<p>Where are the snapshots stored?  Right next to your test in a folder
called <code>snapshots</code> as individual <a href="https://insta.rs/docs/snapshot-files/"><code>.snap</code> files</a>.</p>
<p>Got curious?</p>
<ul>
<li><a href="https://insta.rs/docs/quickstart/">Read the introduction</a></li>
<li><a href="https://insta.rs/docs/">Read the main documentation</a> which does not just
cover the API of the crate but also many of the details of how it works.</li>
<li>There is a screencast that shows the entire workflow: <a href="https://www.youtube.com/watch?v=rCHrMqE4JOY&amp;feature=youtu.be">watch the insta
introduction screencast</a>.</li>
</ul>
<h2 id="writing-tests"><a href="#writing-tests">Writing Tests</a></h2>
<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="kw">use </span>insta::assert_debug_snapshot;

<span class="attr">#[test]
</span><span class="kw">fn </span>test_snapshots() {
    <span class="macro">assert_debug_snapshot!</span>(<span class="macro">vec!</span>[<span class="number">1</span>, <span class="number">2</span>, <span class="number">3</span>]);
}</code></pre></div>
<p>The recommended flow is to run the tests once, have them fail and check
if the result is okay.  By default the new snapshots are stored next
to the old ones with the extra <code>.new</code> extension.  Once you are satisfied
move the new files over.  To simplify this workflow you can use
<code>cargo insta review</code> (requires
<a href="https://crates.io/crates/cargo-insta"><code>cargo-insta</code></a>) which will let you
interactively review them:</p>
<div class="example-wrap"><pre class="language-text"><code>$ cargo test
$ cargo insta review
</code></pre></div><h2 id="assertion-macros"><a href="#assertion-macros">Assertion Macros</a></h2>
<p>This crate exports multiple macros for snapshot testing:</p>
<ul>
<li><a href="macro.assert_snapshot.html" title="macro insta::assert_snapshot"><code>assert_snapshot!</code></a> for comparing basic string snapshots.</li>
<li><a href="macro.assert_debug_snapshot.html" title="macro insta::assert_debug_snapshot"><code>assert_debug_snapshot!</code></a> for comparing <a href="https://doc.rust-lang.org/1.70.0/core/fmt/macros/derive.Debug.html" title="derive core::fmt::macros::Debug"><code>Debug</code></a> outputs of values.</li>
<li><a href="macro.assert_display_snapshot.html" title="macro insta::assert_display_snapshot"><code>assert_display_snapshot!</code></a> for comparing <a href="https://doc.rust-lang.org/1.70.0/core/fmt/trait.Display.html" title="trait core::fmt::Display"><code>Display</code></a> outputs of values.</li>
</ul>
<p>The following macros require the use of serde’s <a href="serde::Serialize"><code>Serialize</code></a>:</p>
<p>For macros that work with [<code>serde</code>] this crate also permits redacting of
partial values.  See <a href="https://insta.rs/docs/redactions/">redactions in the
documentation</a> for more information.</p>
<h2 id="snapshot-updating"><a href="#snapshot-updating">Snapshot updating</a></h2>
<p>During test runs snapshots will be updated according to the <code>INSTA_UPDATE</code>
environment variable.  The default is <code>auto</code> which will write all new
snapshots into <code>.snap.new</code> files if no CI is detected so that
<a href="https://crates.io/crates/cargo-insta"><code>cargo-insta</code></a>
can pick them up.  Normally you don’t have to change this variable.</p>
<p><code>INSTA_UPDATE</code> modes:</p>
<ul>
<li><code>auto</code>: the default. <code>no</code> for CI environments or <code>new</code> otherwise</li>
<li><code>always</code>: overwrites old snapshot files with new ones unasked</li>
<li><code>unseen</code>: behaves like <code>always</code> for new snapshots and <code>new</code> for others</li>
<li><code>new</code>: write new snapshots into <code>.snap.new</code> files</li>
<li><code>no</code>: does not update snapshot files at all (just runs tests)</li>
</ul>
<p>When <code>new</code> or <code>auto</code> is used as mode the <a href="https://crates.io/crates/cargo-insta"><code>cargo-insta</code></a>
command can be used to review the snapshots conveniently:</p>
<div class="example-wrap"><pre class="language-text"><code>$ cargo insta review
</code></pre></div>
<p>“enter” or “a” accepts a new snapshot, “escape” or “r” rejects,
“space” or “s” skips the snapshot for now.</p>
<p>For more information <a href="https://insta.rs/docs/cli/">read the cargo insta docs</a>.</p>
<h2 id="inline-snapshots"><a href="#inline-snapshots">Inline Snapshots</a></h2>
<p>Additionally snapshots can also be stored inline.  In that case the format
for the snapshot macros is <code>assert_snapshot!(reference_value, @&quot;snapshot&quot;)</code>.
The leading at sign (<code>@</code>) indicates that the following string is the
reference value.  <code>cargo-insta</code> will then update that string with the new
value on review.</p>
<p>Example:</p>

<div class="example-wrap ignore"><a href="#" class="tooltip" title="This example is not tested">ⓘ</a><pre class="rust rust-example-rendered"><code><span class="attr">#[derive(Serialize)]
</span><span class="kw">pub struct </span>User {
   username: String,
}

<span class="macro">assert_yaml_snapshot!</span>(User {
   username: <span class="string">&quot;john_doe&quot;</span>.to_string(),
}, @<span class="string">&quot;&quot;</span>);</code></pre></div>
<p>Like with normal snapshots after the initial test failure you can run
<code>cargo insta review</code> to accept the change.  The file will then be updated
automatically.</p>
<h2 id="features"><a href="#features">Features</a></h2>
<p>The following features exist:</p>
<ul>
<li><code>csv</code>: enables CSV support (via serde)</li>
<li><code>json</code>: enables JSON support (via serde)</li>
<li><code>ron</code>: enables RON support (via serde)</li>
<li><code>toml</code>: enables TOML support (via serde)</li>
<li><code>yaml</code>: enables YAML support (via serde)</li>
<li><code>redactions</code>: enables support for redactions</li>
<li><code>filters</code>: enables support for filters</li>
<li><code>glob</code>: enables support for globbing ([<code>glob!</code>])</li>
<li><code>colors</code>: enables color output (enabled by default)</li>
</ul>
<p>For legacy reasons the <code>json</code> and <code>yaml</code> features are enabled by default
in limited capacity.  You will receive a deprecation warning if you are
not opting into them but for now the macros will continue to function.</p>
<p>Enabling any of the serde based formats enables the hidden <code>serde</code> feature
which gates some serde specific APIs such as [<code>Settings::set_info</code>].</p>
<h2 id="dependencies"><a href="#dependencies">Dependencies</a></h2>
<p><code>insta</code> tries to be light in dependencies but this is tricky to accomplish
given what it tries to do.  By default it currently depends on <code>serde</code> for
the [<code>assert_toml_snapshot!</code>] and [<code>assert_yaml_snapshot!</code>] macros.  In
the future this default dependencies will be removed.  To already benefit
from this optimization you can disable the default features and manually
opt into what you want.</p>
<h2 id="settings-1"><a href="#settings-1">Settings</a></h2>
<p>There are some settings that can be changed on a per-thread (and thus
per-test) basis.  For more information see <a href="struct.Settings.html" title="struct insta::Settings">Settings</a>.</p>
<p>Additionally Insta will load a YAML config file with settings that change
the behavior of insta between runs.  It’s loaded from any of the following
locations: <code>.config/insta.yaml</code>, <code>insta.yaml</code> and <code>.insta.yaml</code> from the
workspace root.  The following config options exist:</p>
<div class="example-wrap"><pre class="language-yaml"><code>behavior:
  # also set by INSTA_FORCE_UPDATE
  force_update: true/false
  # also set by INSTA_FORCE_PASS
  force_pass: true/false
  # also set by INSTA_OUTPUT
  output: &quot;diff&quot; | &quot;summary&quot; | &quot;minimal&quot; | &quot;none&quot;
  # also set by INSTA_UPDATE
  update: &quot;auto&quot; | &quot;always&quot; | &quot;new&quot; | &quot;unseen&quot; | &quot;no&quot;
  # also set by INSTA_GLOB_FAIL_FAST
  glob_fail_fast: true/false

# these are used by cargo insta test
test:
  # also set by INSTA_TEST_RUNNER
  runner: &quot;auto&quot; | &quot;cargo-test&quot; | &quot;nextest&quot;
  # automatically assume --review was passed to cargo insta test
  auto_review: true/false
  # automatically assume --accept-unseen was passed to cargo insta test
  auto_accept_unseen: true/false

# these are used by cargo insta review
review:
  # also look for snapshots in ignored folders
  include_ignored: true / false
  # also look for snapshots in hidden folders
  include_hidden: true / false
  # show a warning if undiscovered (ignored or hidden) snapshots are found.
  # defaults to true but creates a performance hit.
  warn_undiscovered: true / false
</code></pre></div><h2 id="optional-faster-runs"><a href="#optional-faster-runs">Optional: Faster Runs</a></h2>
<p>Insta benefits from being compiled in release mode, even as dev dependency.  It
will compile slightly slower once, but use less memory, have faster diffs and
just generally be more fun to use.  To achieve that, opt <code>insta</code> and <code>similar</code>
(the diffing library) into higher optimization in your <code>Cargo.toml</code>:</p>
<div class="example-wrap"><pre class="language-yaml"><code>[profile.dev.package.insta]
opt-level = 3

[profile.dev.package.similar]
opt-level = 3
</code></pre></div></div></details><h2 id="modules" class="small-section-header"><a href="#modules">Modules</a></h2><ul class="item-table"><li><div class="item-name"><a class="mod" href="internals/index.html" title="mod insta::internals">internals</a></div><div class="desc docblock-short">Exposes some library internals.</div></li></ul><h2 id="macros" class="small-section-header"><a href="#macros">Macros</a></h2><ul class="item-table"><li><div class="item-name"><a class="macro" href="macro.allow_duplicates.html" title="macro insta::allow_duplicates">allow_duplicates</a></div><div class="desc docblock-short">Utility macro to permit a multi-snapshot run where all snapshots match.</div></li><li><div class="item-name"><a class="macro" href="macro.assert_debug_snapshot.html" title="macro insta::assert_debug_snapshot">assert_debug_snapshot</a></div><div class="desc docblock-short">Asserts a <code>Debug</code> snapshot.</div></li><li><div class="item-name"><a class="macro" href="macro.assert_display_snapshot.html" title="macro insta::assert_display_snapshot">assert_display_snapshot</a></div><div class="desc docblock-short">Asserts a <code>Display</code> snapshot.</div></li><li><div class="item-name"><a class="macro" href="macro.assert_snapshot.html" title="macro insta::assert_snapshot">assert_snapshot</a></div><div class="desc docblock-short">Asserts a string snapshot.</div></li><li><div class="item-name"><a class="macro" href="macro.with_settings.html" title="macro insta::with_settings">with_settings</a></div><div class="desc docblock-short">Settings configuration macro.</div></li></ul><h2 id="structs" class="small-section-header"><a href="#structs">Structs</a></h2><ul class="item-table"><li><div class="item-name"><a class="struct" href="struct.MetaData.html" title="struct insta::MetaData">MetaData</a></div><div class="desc docblock-short">Snapshot metadata information.</div></li><li><div class="item-name"><a class="struct" href="struct.Settings.html" title="struct insta::Settings">Settings</a></div><div class="desc docblock-short">Configures how insta operates at test time.</div></li><li><div class="item-name"><a class="struct" href="struct.Snapshot.html" title="struct insta::Snapshot">Snapshot</a></div><div class="desc docblock-short">A helper to work with stored snapshots.</div></li></ul></section></div></main></body></html>