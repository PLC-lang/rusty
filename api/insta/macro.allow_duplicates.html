<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Utility macro to permit a multi-snapshot run where all snapshots match."><title>allow_duplicates in insta - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-46f98efaafac5295.ttf.woff2,FiraSans-Regular-018c141bf0843ffd.woff2,FiraSans-Medium-8f9a781e4970d388.woff2,SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2,SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../static.files/rustdoc-492a78a4a87dcc01.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="insta" data-themes="" data-resource-suffix="" data-rustdoc-version="1.82.0 (f6e511eec 2024-10-15)" data-channel="1.82.0" data-search-js="search-a99f1315e7cc5121.js" data-settings-js="settings-4313503d2e1961c2.js" ><script src="../static.files/storage-118b08c4c78b968e.js"></script><script defer src="sidebar-items.js"></script><script defer src="../static.files/main-921df33f47b8780c.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-3b12f09e550e0385.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc macro"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../insta/index.html">insta</a><span class="version">1.40.0</span></h2></div><div class="sidebar-elems"></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><h1>Macro <a href="index.html">insta</a>::<wbr><a class="macro" href="#">allow_duplicates</a><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><span class="out-of-band"><a class="src" href="../src/insta/macros.rs.html#525-531">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><pre class="rust item-decl"><span class="macro">macro_rules!</span> allow_duplicates {
    ($(<span class="macro-nonterminal">$x</span>:tt)<span class="kw-2">*</span>) =&gt; { ... };
}</pre><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Utility macro to permit a multi-snapshot run where all snapshots match.</p>
<p>Within this block, insta will allow an assertion to be run more than once
(even inline) without generating another snapshot.  Instead it will assert
that snapshot expressions visited more than once are matching.</p>

<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="macro">insta::allow_duplicates!</span> {
    <span class="kw">for </span>x <span class="kw">in </span>(<span class="number">0</span>..<span class="number">10</span>).step_by(<span class="number">2</span>) {
        <span class="kw">let </span>is_even = x % <span class="number">2 </span>== <span class="number">0</span>;
        <span class="macro">insta::assert_debug_snapshot!</span>(is_even, @<span class="string">"true"</span>);
    }
}</code></pre></div>
<p>The first snapshot assertion will be used as a gold master and every further
assertion will be checked against it.  If they don’t match the assertion will
fail.</p>
</div></details></section></div></main></body></html>