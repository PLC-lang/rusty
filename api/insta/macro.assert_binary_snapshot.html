<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="(Experimental) Asserts a binary snapshot in the form of a `Vec&#60;u8&#62;`."><title>assert_binary_snapshot in insta - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-46f98efaafac5295.ttf.woff2,FiraSans-Regular-018c141bf0843ffd.woff2,FiraSans-Medium-8f9a781e4970d388.woff2,SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2,SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../static.files/rustdoc-b0742ba02757f159.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="insta" data-themes="" data-resource-suffix="" data-rustdoc-version="1.83.0 (90b35a623 2024-11-26)" data-channel="1.83.0" data-search-js="search-f0d225181b97f9a4.js" data-settings-js="settings-805db61a62df4bd2.js" ><script src="../static.files/storage-1d39b6787ed640ff.js"></script><script defer src="sidebar-items.js"></script><script defer src="../static.files/main-f070b9041d14864c.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-0111fcff984fae8f.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc macro"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../insta/index.html">insta</a><span class="version">1.42.0</span></h2></div><div class="sidebar-elems"><div id="rustdoc-modnav"></div></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><span class="rustdoc-breadcrumbs"><a href="index.html">insta</a></span><h1>Macro <span class="macro">assert_binary_snapshot</span><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../src/insta/macros.rs.html#398-419">source</a> </span></div><pre class="rust item-decl"><span class="macro">macro_rules!</span> assert_binary_snapshot {
    (<span class="macro-nonterminal">$name_and_extension</span>:expr, <span class="macro-nonterminal">$value</span>:expr $(,)<span class="question-mark">?</span>) =&gt; { ... };
    (<span class="macro-nonterminal">$name_and_extension</span>:expr, <span class="macro-nonterminal">$value</span>:expr, <span class="macro-nonterminal">$debug_expr</span>:expr $(,)<span class="question-mark">?</span>) =&gt; { ... };
}</pre><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>(Experimental)
Asserts a binary snapshot in the form of a <a href="https://doc.rust-lang.org/1.83.0/alloc/vec/struct.Vec.html" title="struct alloc::vec::Vec"><code>Vec&lt;u8&gt;</code></a>.</p>
<p>The contents get stored in a separate file next to the metadata file. The extension for this
file must be passed as part of the name. For an implicit snapshot name just an extension can be
passed starting with a <code>.</code>.</p>
<p>This feature is considered experimental: we may make incompatible changes for the next couple
of versions after 1.41.</p>
<p>Examples:</p>

<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="comment">// implicit name:
</span><span class="macro">insta::assert_binary_snapshot!</span>(<span class="string">".txt"</span>, <span class="string">b"abcd"</span>.to_vec());

<span class="comment">// named:
</span><span class="macro">insta::assert_binary_snapshot!</span>(<span class="string">"my_snapshot.bin"</span>, [<span class="number">0</span>, <span class="number">1</span>, <span class="number">2</span>, <span class="number">3</span>].to_vec());</code></pre></div>
</div></details></section></div></main></body></html>