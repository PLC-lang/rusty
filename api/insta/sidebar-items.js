window.SIDEBAR_ITEMS = {"macro":["allow_duplicates","assert_debug_snapshot","assert_display_snapshot","assert_snapshot","with_settings"],"mod":["internals"],"struct":["MetaData","Settings","Snapshot"]};