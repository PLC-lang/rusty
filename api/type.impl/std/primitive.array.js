(function() {var type_impls = {
"uuid":[]
};if (window.register_type_impls) {window.register_type_impls(type_impls);} else {window.pending_type_impls = type_impls;}})()