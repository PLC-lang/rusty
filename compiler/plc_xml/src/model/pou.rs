use std::{collections::HashMap, str::FromStr};

use quick_xml::events::Event;

use crate::{error::Error, extensions::GetOrErr, reader::PeekableReader, xml_parser::Parseable};

use super::{action::Action, body::Body, interface::Interface};

#[derive(Debug, Default)]
pub(crate) struct Pou {
    pub name: String,
    pub pou_type: PouType,
    pub body: Body,
    pub actions: Vec<Action>,
    pub interface: Option<Interface>,
}

impl Pou {
    fn with_attributes(self, attributes: HashMap<String, String>) -> Result<Self, Error> {
        Ok(Pou {
            name: attributes.get_or_err("name")?,
            pou_type: attributes.get_or_err("pouType").map(|it| it.parse())??,
            body: self.body,
            actions: self.actions,
            interface: self.interface,
        })
    }
}

#[derive(Debug, Default, Clone, Copy)]
pub(crate) enum PouType {
    #[default]
    Program,
    Function,
    FunctionBlock,
}

impl std::fmt::Display for PouType {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            PouType::Program => write!(f, "PROGRAM"),
            PouType::Function => write!(f, "FUNCTION"),
            PouType::FunctionBlock => write!(f, "FUNCTION_BLOCK"),
        }
    }
}

impl TryFrom<&str> for PouType {
    type Error = Error;

    fn try_from(value: &str) -> Result<Self, Self::Error> {
        match value {
            "program" => Ok(PouType::Program),
            "function" => Ok(PouType::Function),
            "functionBlock" => Ok(PouType::FunctionBlock),
            _ => Err(Error::UnexpectedElement(value.to_string())),
        }
    }
}

impl Parseable for Pou {
    type Item = Self;

    fn visit(reader: &mut PeekableReader) -> Result<Self::Item, Error> {
        let mut pou = Pou::default().with_attributes(reader.attributes()?)?;
        loop {
            match reader.peek()? {
                Event::Start(tag) => match tag.name().as_ref() {
                    b"interface" => {
                        // XXX: this is very specific to our own xml schema, but does not adhere to the plc open standard
                        reader.consume_until_start(b"content")?;
                        match reader.next()? {
                            Event::Start(tag) => {
                                pou.interface = Some(Interface::new(&reader.read_text(tag.name())?))
                            }
                            _ => reader.consume()?,
                        }
                    }
                    b"body" => {
                        pou.body = Body::visit(reader)?;
                        if let Some(interface) = pou.interface {
                            pou.interface = Some(interface.append_end_keyword(&pou.pou_type));
                        }

                        // TODO: change in order to parse INTERFACE, ACTION etc..
                        reader.consume_until(vec![b"pou"])?;
                        return Ok(pou);
                    }

                    _ => reader.consume()?,
                },

                Event::End(tag) if tag.name().as_ref() == b"pou" => return Ok(pou),

                _ => reader.consume()?,
            }
        }
    }
}

impl FromStr for PouType {
    type Err = Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "program" => Ok(PouType::Program),
            "function" => Ok(PouType::Function),
            "functionBlock" => Ok(PouType::FunctionBlock),
            _ => Err(Error::UnexpectedElement(s.to_string())),
        }
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;

    use crate::{
        model::{pou::Pou, project::Project},
        reader::PeekableReader,
        serializer::{
            XAddData, XBody, XContent, XData, XFbd, XInterface, XLocalVars, XPou, XTextDeclaration,
        },
        xml_parser::Parseable,
    };

    #[test]
    fn empty() {
        let content = XPou::new()
            .with_attribute("xmlns", "http://www.plcopen.org/xml/tc6_0201")
            .with_attribute("name", "foo")
            .with_attribute("pouType", "program")
            .with_interface(
                XInterface::new().with_local_vars(XLocalVars::new().close()).with_add_data(
                    XAddData::new().with_data_data(
                        XData::new()
                            .with_attribute("name", "www.bachmann.at/plc/plcopenxml")
                            .with_attribute("handleUnknown", "implementation")
                            .with_text_declaration(XTextDeclaration::new().with_content(
                                XContent::new().with_data(
                                    r#"
PROGRAM foo
VAR

END_VAR
                    "#,
                                ),
                            )),
                    ),
                ),
            )
            .with_body(XBody::new().with_fbd(XFbd::new().close()))
            .serialize();

        let mut reader = PeekableReader::new(&content);
        assert_debug_snapshot!(Project::pou_entry(&mut reader));
    }

    #[test]
    fn poutype_program() {
        let content =
            XPou::new().with_attribute("name", "foo").with_attribute("pouType", "program").serialize();

        let mut reader = PeekableReader::new(&content);
        assert_debug_snapshot!(Pou::visit(&mut reader));
    }

    #[test]
    fn poutype_function() {
        let content =
            XPou::new().with_attribute("name", "foo").with_attribute("pouType", "function").serialize();

        let mut reader = PeekableReader::new(&content);
        assert_debug_snapshot!(Pou::visit(&mut reader));
    }

    #[test]
    fn poutype_function_block() {
        let content =
            XPou::new().with_attribute("name", "foo").with_attribute("pouType", "functionBlock").serialize();

        let mut reader = PeekableReader::new(&content);
        assert_debug_snapshot!(Pou::visit(&mut reader));
    }

    #[test]
    fn poutype_unknown() {
        let content =
            XPou::new().with_attribute("name", "foo").with_attribute("pouType", "asdasd").serialize();

        let mut reader = PeekableReader::new(&content);
        assert_eq!(
            Pou::visit(&mut reader).unwrap_err().to_string(),
            "Found an unexpected element 'asdasd'".to_string()
        );
    }
}
