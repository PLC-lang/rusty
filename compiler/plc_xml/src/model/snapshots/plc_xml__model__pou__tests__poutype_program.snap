---
source: compiler/plc_xml/src/model/pou.rs
expression: "Pou::visit(&mut reader)"
---
Ok(
    Pou {
        name: "foo",
        pou_type: Program,
        body: Body {
            function_block_diagram: None,
        },
        actions: [],
        interface: None,
    },
)
